Reference/Security/two-factor-authentication/index.md

```diff
@@ -0,0 +1,189 @@
+            identityBuilder.AddTwoFactorProvider<UmbracoAppAuthenticator>(UmbracoAppAuthenticator.Name);
+        }
+    }
+}
+```
+
+
+At this point the 2FA is actually active, but no members have setup 2FA yet.
+The setup of 2FA depends on the type. In the case of App Authenticator, we will add the following to our view showing the edit profile of the member.
+
+```
+@using Umbraco.Cms.Core.Services
+@using Umbraco.Cms.Web.Website.Controllers
+@using Umbraco.Cms.Web.Website.Models
+@using Umbraco9  @* Or whatever your namespace with the QrCodeSetupData model is *@
+@inject MemberModelBuilderFactory memberModelBuilderFactory;
+@inject ITwoFactorLoginService twoFactorLoginService
+@{
+    // Build a profile model to edit
+    var profileModel = await memberModelBuilderFactory
+        .CreateProfileModel()
+        .BuildForCurrentMemberAsync();
+
+    // Show all two factor providers
+    var providerNames = twoFactorLoginService.GetAllProviderNames();
+    if (providerNames.Any())
+    {
+        <div asp-validation-summary="All" class="text-danger"></div>
+        foreach (var providerName in providerNames)
+        {
+            var setupData = await twoFactorLoginService.GetSetupInfoAsync(profileModel.Key, providerName);
+            if (setupData is null)
+            {
+                @using (Html.BeginUmbracoForm<UmbTwoFactorLoginController>(nameof(UmbTwoFactorLoginController.Disable)))
+                {
+                    <input type="hidden" name="providerName" value="@providerName"/>
+                    <button type="submit">Disable @providerName</button>
+                }
+            }
+            else if(setupData is QrCodeSetupData qrCodeSetupData)
+            {
+                @using (Html.BeginUmbracoForm<UmbTwoFactorLoginController>(nameof(UmbTwoFactorLoginController.ValidateAndSaveSetup)))
+                {
+                    <h3>Setup @providerName</h3>
+                    <img src="@qrCodeSetupData.SetupCode.QrCodeSetupImageUrl"/>
+                    <p>Scan the code above with your authenticator app <br /> and enter the resulting code here to validate:</p>
+                    <input type="hidden" name="providerName" value="@providerName"  />
+                    <input type="hidden" name="secret" value="@qrCodeSetupData.Secret"  />
+                    <input type="text" name="code"  />
+                    <button type="submit">Validate & save</button>
+                }
+            }
+        }
+    }
+}
+```
+
+In this razor-code sample, we get the current members unique key and list all registered `ITwoFactorProvider` implementations.
+
+If the `setupData` is `null` for the specified `providerName` it means the provider is already setup. In this case we show a disable button. Otherwise we check the type and show the UI for how to setup the App Authenticator, by showing the QR Code and a input field to validate the code from the App Authenticator.
```

Comment:
```suggestion
In this razor-code sample, we get the current members unique key and list all registered `ITwoFactorProvider` implementations.

If the `setupData` is `null` for the specified `providerName` it means the provider is already set up. In this case, we show a disable button. Otherwise, we check the type and show the UI for how to set up the App Authenticator, by showing the QR Code and an input field to validate the code from the App Authenticator.
```